### PR TITLE
HTTPCLIENT-2284: internal cache storage improvements

### PR DESCRIPTION
**Pull request by ok2c**
Added internal cache storage implementation that can be used to implement custom cache storage backends with specific cache entry eviction strategy.

@arturobernalg Please review.

### CODE REVIEW

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/InternalCacheStorage.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import org.apache.hc.core5.annotation.Internal;
+
+@Internal
+final public class InternalCacheStorage {
+
+    private final Map<String, HttpCacheEntry> map;
+    private final Queue<HttpCacheStorageEntry> evictionQueue;
+    private final Consumer<HttpCacheStorageEntry> evictionCallback;
+
+    public InternalCacheStorage(final int maxEntries, final Consumer<HttpCacheStorageEntry> evictionCallback) {
```
Comment (arturobernalg) [inline httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/InternalCacheStorage.java:46]: shouldn't we check if `maxEntries`is positives?

**Review by arturobernalg — APPROVED** [review]
@ok2c  the change-set looks good. I just add a small question.

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/InternalCacheStorage.java (hunk 1 above)
Comment (ok2c) [reply httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/InternalCacheStorage.java:46]: @arturobernalg I would say not. Not at this level. Input validation should be done at the public API level,

Comment (arturobernalg) [reply httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/InternalCacheStorage.java:46]: fair enough

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/InternalCacheStorage.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        this.evictionCallback = evictionCallback;
+        this.map = new LinkedHashMap<String, HttpCacheEntry>(20, 0.75f, true) {
+
+            @Override
+            protected boolean removeEldestEntry(final Map.Entry<String, HttpCacheEntry> eldest) {
+                if (evictionCallback != null && size() > maxEntries) {
```
Comment (lacompas) [inline httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/InternalCacheStorage.java:52]: This is a change in behaviour, now the maxEntries size limit is no longer honoured by default

**Pull request merged**